File: src/main/java/io/github/chaosdave34/kitpvp/listener/GamePlayerDeathListener.java
package io.github.chaosdave34.kitpvp.listener;

import io.github.chaosdave34.ghutils.utils.PDCUtils;
import io.github.chaosdave34.kitpvp.ExtendedPlayer;
import io.github.chaosdave34.kitpvp.KitPvp;
import io.github.chaosdave34.kitpvp.damagetype.DamageTypes;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import org.bukkit.Bukkit;
import org.bukkit.damage.DamageSource;
import org.bukkit.damage.DamageType;
import org.bukkit.entity.*;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.PlayerDeathEvent;

import java.util.UUID;

public class GamePlayerDeathListener implements Listener {
    @EventHandler
    public void onDeath(PlayerDeathEvent e) {
        Player p = e.getPlayer();
        ExtendedPlayer extendedPlayer = ExtendedPlayer.from(p);
        e.setKeepInventory(true);

        if (extendedPlayer.inGame()) {

            extendedPlayer.incrementTotalDeaths();

            Bukkit.getScheduler().runTaskLater(KitPvp.INSTANCE, () -> extendedPlayer.spawn(), 1);

            e.setCancelled(true);

            // Death Messages
            String name = p.getName();
            Component deathMessage = e.deathMessage();
            String message = deathMessage == null ? "" : PlainTextComponentSerializer.plainText().serialize(deathMessage);

            EntityDamageEvent lastDamageEvent = p.getLastDamageCause();
            if (lastDamageEvent != null) {
                DamageSource damageSource = lastDamageEvent.getDamageSource();
                DamageType damageType = damageSource.getDamageType();

                // Damage by Entity
                if (lastDamageEvent instanceof EntityDamageByEntityEvent lastDamageByEntityEvent) {
                    Entity damager = lastDamageByEntityEvent.getDamager();

                    // Kill Effect
                    if (damager instanceof Player killer) {
                        ExtendedPlayer.from(killer).killedPlayer(p);
                    } else if (damager instanceof Projectile projectile && projectile.getShooter() instanceof Player killer) {
                        ExtendedPlayer.from(killer).killedPlayer(p);
                    }

                    // Lightning
                    else if (damager instanceof LightningStrike lightningStrike) {
                        if (lightningStrike.getCausingPlayer() != null) {
                            Player killer = lightningStrike.getCausingPlayer();
                            message = name + " was killed by " + killer.getName() + " with lightning";

                            ExtendedPlayer.from(killer).killedPlayer(p);
                            KitPvp.INSTANCE.getCosmeticHandler().triggerKillEffect(killer, p);
                        }
                    }
                    // Turret
                    if (damager instanceof Firework firework) {
                        if (firework.getShooter() instanceof Husk husk) {
                            UUID turretOwnerUUUID = PDCUtils.getOwner(husk);
                            if (turretOwnerUUUID != null) {
                                Player turretOwner = Bukkit.getPlayer(turretOwnerUUUID);
                                if (turretOwner != null) {
                                    message = name + " was killed by " + turretOwner.getName() + "'s turret";

                                    ExtendedPlayer.from(turretOwner).killedPlayer(p);
                                    KitPvp.INSTANCE.getCosmeticHandler().triggerKillEffect(turretOwner, p);
                                }
                            }
                        }
                    }
                    // Companion
                    else if (damager.hasMetadata("companion")) {
                        UUID ownerUUID = PDCUtils.getOwner(damager);
                        if (ownerUUID != null) {
                            Player companionOwner = Bukkit.getPlayer(ownerUUID);
                            if (companionOwner != null) {
                                message = name + " was killed by " + companionOwner.getName() + "'s companion";

                                ExtendedPlayer.from(companionOwner).killedPlayer(p);
                                KitPvp.INSTANCE.getCosmeticHandler().triggerKillEffect(companionOwner, p);
                            }
                        }

                    }
                } else if (damageType == DamageTypes.LAND) {
                    message = name + " tried to land";
                } else if (damageType == DamageTypes.ESCAPE) {
                    message = name + " tried to escape";
                }

                Bukkit.broadcast(Component.text("☠ ", NamedTextColor.RED)
                        .append(Component.text(message, NamedTextColor.GRAY)));
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/listener/GamePlayerDeathListener.kt
package io.github.chaosdave34.kitpvp.listener

import io.github.chaosdave34.ghutils.utils.PDCUtils
import io.github.chaosdave34.kitpvp.ExtendedPlayer
import io.github.chaosdave34.kitpvp.KitPvp
import io.github.chaosdave34.kitpvp.damagetype.DamageTypes
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import org.bukkit.Bukkit
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.PlayerDeathEvent

class GamePlayerDeathListener : Listener {
    @EventHandler
    fun onDeath(event: PlayerDeathEvent) {
        val player = event.player
        val extendedPlayer = ExtendedPlayer.from(player)
        event.keepInventory = true

        if (extendedPlayer.inGame()) {
            extendedPlayer.incrementTotalDeaths()

            Bukkit.getScheduler().runTaskLater(KitPvp.INSTANCE, Runnable { extendedPlayer.spawn() }, 1)

            event.isCancelled = true

            // Death Messages
            val name = player.name
            val deathMessage = event.deathMessage()
            var message = if (deathMessage == null) "" else PlainTextComponentSerializer.plainText().serialize(deathMessage)

            val lastDamageEvent = player.lastDamageCause ?: return
            val damageSource = lastDamageEvent.damageSource
            val damageType = damageSource.damageType

            // Damage by Entity
            if (lastDamageEvent is EntityDamageByEntityEvent) {
                val damager: Entity = lastDamageEvent.damager

                // Kill Credits
                if (damager is Player) {
                    ExtendedPlayer.from(damager).killedPlayer(player)
                } else if (damager is Projectile && damager.shooter is Player) {
                    val killer = damager.shooter as Player
                    ExtendedPlayer.from(killer).killedPlayer(player)
                }

                // Firework
                if (damager is LightningStrike) {
                    val killer = damager.causingPlayer

                    if (killer != null) {
                        message = "$name was killed by ${killer.name} with lightning"
                        ExtendedPlayer.from(killer).killedPlayer(player)
                    }
                }
                // Turret
                else if (damager is Firework) {
                    val shooter = damager.shooter
                    if (shooter is Husk) {
                        val turretOwnerUUUID = PDCUtils.getOwner(shooter)
                        if (turretOwnerUUUID != null) {
                            val turretOwner = Bukkit.getPlayer(turretOwnerUUUID)
                            if (turretOwner != null) {
                                message = "$name was killed by ${turretOwner.name}'s turret"
                                ExtendedPlayer.from(turretOwner).killedPlayer(player)
                            }
                        }
                    }
                    // Companions
                } else if (damager.hasMetadata("companion")) {
                    val ownerUUID = PDCUtils.getOwner(damager)
                    if (ownerUUID != null) {
                        val companionOwner = Bukkit.getPlayer(ownerUUID)
                        if (companionOwner != null) {
                            message = "$name was killed by ${companionOwner.name}'s companion"
                            ExtendedPlayer.from(companionOwner).killedPlayer(player)
                        }
                    }
                }

                // Missing death messages
                // Magic Wand / Magic damage by item
                if (lastDamageEvent.cause == EntityDamageEvent.DamageCause.MAGIC)
                    message = "$name was killed by ${damager.name} using magic"
                else if (lastDamageEvent.cause == EntityDamageEvent.DamageCause.DROWNING)
                    message = "$name was drowned by ${damager.name}"

                if (damageType == DamageTypes.BLACK_HOLE)
                    message = "$name was killed by ${damager.name}'s black hole"

            } else if (damageType === DamageTypes.LAND)
                message = "$name tried to land"
            else if (damageType === DamageTypes.ESCAPE)
                message = "$name tried to escape"


            Bukkit.broadcast(
                Component.text("☠ ", NamedTextColor.RED)
                    .append(Component.text(message, NamedTextColor.GRAY))
            )
        }
    }
}